### DISCUSSION TITLE
Installer failed Pi 3B running Venus OS

### DISCUSSION THREAD

**Discussion by tgsail1** [CharlesGodwin/pymagnum#53] [General]
Hi tried installing on a PI3B running Venus OS (stripped out linux). The idea is to integrate magnum inverter/charger into victron/venus environment. PIP3 is installed and runs. Got errors and noticed the install script calls pip. Set alias pip=pip3. Got further into the install but still crashed. Here is the error dump. Maybe you can shed light on where the error is actually occurring. Note: Venus doesn't support apt, but does use opkg. Thanks.

root@raspberrypi2:/# pip3 install --upgrade pymagnum
Collecting pymagnum
  Using cached pymagnum-2.0.2-py3-none-any.whl (22 kB)
Collecting tzlocal
  Using cached tzlocal-5.1-py3-none-any.whl (21 kB)
Requirement already satisfied, skipping upgrade: pyserial in /usr/lib/python3.8/site-packages (from pymagnum) (3.4)
Collecting uptime
  Using cached uptime-3.0.1.tar.gz (6.6 kB)
Collecting backports.zoneinfo; python_version < "3.9"
  Using cached backports.zoneinfo-0.2.1.tar.gz (74 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
ERROR: Exception:
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/pip/_internal/cli/base_command.py", line 186, in _main
    status = self.run(options, args)
  File "/usr/lib/python3.8/site-packages/pip/_internal/commands/install.py", line 331, in run
    resolver.resolve(requirement_set)
  File "/usr/lib/python3.8/site-packages/pip/_internal/legacy_resolve.py", line 177, in resolve
    discovered_reqs.extend(self._resolve_one(requirement_set, req))
  File "/usr/lib/python3.8/site-packages/pip/_internal/legacy_resolve.py", line 333, in _resolve_one
    abstract_dist = self._get_abstract_dist_for(req_to_install)
  File "/usr/lib/python3.8/site-packages/pip/_internal/legacy_resolve.py", line 282, in _get_abstract_dist_for
    abstract_dist = self.preparer.prepare_linked_requirement(req)
  File "/usr/lib/python3.8/site-packages/pip/_internal/operations/prepare.py", line 515, in prepare_linked_requirement
    abstract_dist = _get_prepared_distribution(
  File "/usr/lib/python3.8/site-packages/pip/_internal/operations/prepare.py", line 95, in _get_prepared_distribution
    abstract_dist.prepare_distribution_metadata(finder, build_isolation)
  File "/usr/lib/python3.8/site-packages/pip/_internal/distributions/sdist.py", line 38, in prepare_distribution_metadata
    self._setup_isolation(finder)
  File "/usr/lib/python3.8/site-packages/pip/_internal/distributions/sdist.py", line 96, in _setup_isolation
    reqs = backend.get_requires_for_build_wheel()
  File "/usr/lib/python3.8/site-packages/pip/_vendor/pep517/wrappers.py", line 151, in get_requires_for_build_wheel
    return self._call_hook('get_requires_for_build_wheel', {
  File "/usr/lib/python3.8/site-packages/pip/_vendor/pep517/wrappers.py", line 255, in _call_hook
    raise BackendUnavailable(data.get('traceback', ''))
pip._vendor.pep517.wrappers.BackendUnavailable: Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py", line 63, in _build_backend
    obj = import_module(mod_path)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/var/volatile/tmp/pip-build-env-q1q0wfaf/overlay/lib/python3.8/site-packages/setuptools/__init__.py", line 16, in <module>
    from .dist import Distribution
  File "/var/volatile/tmp/pip-build-env-q1q0wfaf/overlay/lib/python3.8/site-packages/setuptools/dist.py", line 27, in <module>
    from .extern.packaging.markers import InvalidMarker, Marker
  File "/var/volatile/tmp/pip-build-env-q1q0wfaf/overlay/lib/python3.8/site-packages/setuptools/_vendor/packaging/markers.py", line 11, in <module>
    from ._parser import (
  File "/var/volatile/tmp/pip-build-env-q1q0wfaf/overlay/lib/python3.8/site-packages/setuptools/_vendor/packaging/_parser.py", line 10, in <module>
    from ._tokenizer import DEFAULT_RULES, Tokenizer
  File "/var/volatile/tmp/pip-build-env-q1q0wfaf/overlay/lib/python3.8/site-packages/setuptools/_vendor/packaging/_tokenizer.py", line 3, in <module>
    from dataclasses import dataclass
ModuleNotFoundError: No module named 'dataclasses'

**Comment by CharlesGodwin** [comment]
Hmmm. I'll investigate.

**Comment by tgsail1** [comment]
Thanks

**Comment by CharlesGodwin** [comment]
I notice you aren't using `sudo`. I don't know if that's an issue with Venus OS but you could try `sudo pip3 install pymagnum --upgrade`.

If that doesn't work try running this  `pip3 install dataclasses` and then try again. If that works, I'll amend the package.

**Reply by CharlesGodwin** [reply]
dataclasses was supposed to be part of  python 3.7 and you're running 3.8 so it shouldn't be a problem. As I have no access to Venus OS all I can recommend is you download / clone the source code in the git repository and try to build it locally. If that fails, talk to Venus OS support

**Comment by tgsail1** [comment]
On venus os sudo doesn’t work. Root has all privileges so it shouldn’t be necessary. Installed dataclasses, but the error is still there. Seems like this has to be a path problem or possibly a python version issue
T

**Comment by tgsail1** [comment]
Thanks. Dataclasses is installed and is accepted by python in interactive mode, so it know it is there. There must be a path problem in the build process. Will try the source code.

**Comment by tgsail1** [comment]
The error appears to be upon installing of tzlocal if that is any help. Tried "pip3 install tzlocal" and it generates the same error.

**Reply by CharlesGodwin** [reply]
That's helpful as I think I can get rid of tzlocal. But not today I'm out for the next 4-5 hours.  If no news in 24 hours remind me

**Comment by CharlesGodwin** [comment]
I have built a new version that does not use tzlocal. for now it is branch `Release-2.0.3` and has not been merged into the main branch.
To test it:

- Clone or unzip branch named Release-2.0.3 and build a test version. 
- Use script `build-draft.sh` to build a file name `./dist/pymagnum-2.0.3.zip`
- install that file with `pip install ./dist/pymagnum-2.0.3.zip`
- Report result

If this works, I'll publish  an official update in the next few days.

**Comment by tgsail1** [comment]
Installed using build-draft. Got some errors but it created the zip file. ran pip and it claims to have installed. Will see how it works. Might be a few days before I can test it on the magnasine. Thanks for your help
T

**Reply by CharlesGodwin** [reply]
Thanks. Do you remember the errors? Alas, while testing I ran into Python 3.11 and 3.12 that have introduced new rules for `pip` installs. I need to review my install instructions and how I build the package before I release an official release. With luck I'll be done when you have tested.
I'm curious what you packets look like with just an inverter attached. Or do you have other bits? If you wish, generate a log file and post it. Add `--log` to a `magtest` run. `magtest` will generate a log file and display a message like:
`Output was logged to D:\users\charl\Documents\projects\pymagnum\magtest_2023-10-16T14-22-32.txt`

**Comment by tgsail1** [comment]
say, does write_packets.py by any chance make a dummy magnasine on an RS-485 network?

**Reply by CharlesGodwin** [reply]
Yes and No. `write-packets.py` writes Magnum Energy compatible packets to a RS485/USB dongle or HAT. But **NO,** it is not meant to write packets to a real, populated Magnum Energy network. **Do not run this code on a machine connected to a real Magnum Energy network. This code is used to test reading packets.** To use it you need two Pi, each with a RS485 dongle or HAT, wired together. One Pi will run this code and use as its data, the contents of a file generated on a real system using `magtest`. The output from `magtest` looks like this
<pre>Length:21 REMOTE_00 =>400001EC000976B0010033152C1F73010003025800
Length:21 REMOTE_00 =>0005DA8A643C200600D2FF88001400000000000000
Length:21 REMOTE_00 =>400001EC000976B0010033152C1F73010003025800
Length:21 REMOTE_80 =>0005DA8A643C200600D2FF88001401130000B50080
Length:18 BMK_81    =>8155132AFF3812F615FBFEEE125200040A01</pre>
The value to the right of ==> is hex and is written out to the serial port.

Another simpler way to test / debug code processing packets is to define a dummy / mock reader in any of the example programs. If the `--device` name is prefixed with a `!` the rest of the name is treated as a filename and is read for data. The format of the text
        must be the same as the output generated by the ``magtest`` program.

**Comment by tgsail1** [comment]
Charles- Thanks. Got the programs running on 2 rpis . magtest gives a no network found error most of the time, but about every 4th time it reads something. It seems to be a little more responsive if I set the timeout on write to 0.005 and magtest to 0.010.
Here is an example of ok? data coming through. Might have a flaky adapter. I don’t recall what the build error was, but magdump does not run.
Here is what magtest prints:
***@***.***:~# magtest -d /dev/ttyUSB0 --timeout .010
Magnum Test Version:2.0.3
Options:cleanpackets=True, device='/dev/ttyUSB0', log=False, packets=50, timeout=0.01, trace=False
Length:21 REMOTE_A0 =>00003C04500F170601C8A5860100073714006E00A0
Length:15 UNKNOWN   =>0103000B0020C80103000B0001F5C8
Length:16 REMOTE_C  =>0103000B0001F5C80203000B0001F5FB
Length:37 UNKNOWN   =>00003C04500F170601C8A58601001C1C87461D14A10103000B0001F5C80203000B0001F5FB
Length:21 UNKNOWN   =>A102343A007F0103000B0001F5C80203000B0001F5
Length:37 UNKNOWN   =>0103000B0001F5C8400000F60002780001003311241E6B0000010258000203000B0001F5FB
Length: 3 UNKNOWN   =>0141FE
Length:21 INVERTER  =>400000F60002770001003311241E6B000001025800
Length:21 REMOTE_80 =>00003C04500F170601C8A5860100000000002B0080
Length:21 INVERTER  =>400000F60002780001003311241E6B000001025800
Length:21 REMOTE_00 =>00003C04500F170601C8A586010000000000000000
Length:21 INVERTER  =>400000F60002780001003311241E6B000001025800
Length:21 REMOTE_A0 =>00003C04500F170601C8A5860100073714006E00A0
Length: 6 AGS_A1    =>A102343A007C
Length:21 INVERTER  =>400000F60002780001003311241E6B000001025800
Length:21 REMOTE_A1 =>00003C04500F170601C8A58601001C1C87461D14A1
Length: 6 AGS_A1    =>A102343A007C
Length:21 INVERTER  =>400000F60002780001003311241E6B000001025800
Length:21 REMOTE_A2 =>00003C04500F170601C8A5860100465000781478A2
Length:18 BMK_81    =>81550991FFBA089B0C77FFBFE11100390A01
Length:21 INVERTER  =>400000F60002780001003311241E6B000001025800
Length:21 REMOTE_A3 =>00003C04500F170601C8A58601005E5F00240200A3
Length:21 INVERTER  =>400000F60002780001003311241E6B000001025800
Length:21 REMOTE_A4 =>00003C04500F170601C8A58601001E1E00000000A4
Length: 6 AGS_A1    =>A102343A007A
Length:21 INVERTER  =>400000F60002780001003311241E6B000001025800
Length:21 REMOTE_80 =>00003C04500F170601C8A5860100000000002B0080
Length:21 INVERTER  =>400000F60002780001003311241E6B000001025800
Length:21 REMOTE_00 =>00003C04500F170601C8A586010000000000000000
Length:21 INVERTER  =>400000F60002780001003311241E6B000001025800
Length:21 REMOTE_A0 =>00003C04500F170601C8A5860100073714006E00A0
Length: 6 AGS_A1    =>A102343A007A
Length:21 INVERTER  =>400000F60002790001003311241E6B000001025800
Length:21 REMOTE_A1 =>00003C04500F170601C8A58601001C1C87461D14A1
Length: 6 AGS_A1    =>A102343C007B
Length:21 INVERTER  =>400000F60002790001003311241E6B000001025800
Length:21 REMOTE_A2 =>00003C04500F170601C8A5860100465000781478A2
Length:18 BMK_81    =>81550991FFBA089B0C77FFBFE11100390A01
Length:21 INVERTER  =>400000F60002790001003311241E6B000001025800
Length:21 REMOTE_A3 =>00003C04500F170601C8A58601005E5F00240200A3
Length:21 INVERTER  =>400000F60002790001003311241E6B000001025800
Length:21 REMOTE_A4 =>00003C04500F170601C8A58601001E1E00000000A4
Length: 6 AGS_A1    =>A102343A007A
Packets:43 of 50 with 5 UNKNOWN, in 5.46 seconds
INVERTER Detected
REMOTE Detected
RTR not connected
BMK Detected
AGS Detected
PT100 not supported
ACLD not supported

This is useful for debugging everything else on the victron side before taking everything to the boat.
T

**Comment by tgsail1** [comment]
Resolved the issue. Needed terminating resistors on the RS485 lines. I didn’t have 120 ohm resistors handy, but 500 ohm improved the results dramatically. I think I’m ready to move on to the dbus writing.

**Comment by tgsail1** [comment]
Here is the install error. Interestingly magtest runs, but if I try to import magnum into python it fails (ACLDDevice not found, magnum is not a package).:
***@***.***:~/pymagnum-8df2de5243ad56e453096926d7605a8861a1d54f# sh build_draft.sh
'nvalid command name 'build
running sdist
running egg_info
writing pymagnum.egg-info/PKG-INFO
writing dependency_links to pymagnum.egg-info/dependency_links.txt
writing entry points to pymagnum.egg-info/entry_points.txt
writing requirements to pymagnum.egg-info/requires.txt
writing top-level names to pymagnum.egg-info/top_level.txt
reading manifest file 'pymagnum.egg-info/SOURCES.txt'
writing manifest file 'pymagnum.egg-info/SOURCES.txt'
running check
copying files to pymagnum-2.0.3...
copying pymagnum.egg-info/PKG-INFO -> pymagnum-2.0.3/pymagnum.egg-info
copying pymagnum.egg-info/SOURCES.txt -> pymagnum-2.0.3/pymagnum.egg-info
copying pymagnum.egg-info/dependency_links.txt -> pymagnum-2.0.3/pymagnum.egg-info
copying pymagnum.egg-info/entry_points.txt -> pymagnum-2.0.3/pymagnum.egg-info
copying pymagnum.egg-info/requires.txt -> pymagnum-2.0.3/pymagnum.egg-info
copying pymagnum.egg-info/top_level.txt -> pymagnum-2.0.3/pymagnum.egg-info
Writing pymagnum-2.0.3/setup.cfg
creating 'dist/pymagnum-2.0.3.zip' and adding 'pymagnum-2.0.3' to it
adding 'pymagnum-2.0.3'
adding 'pymagnum-2.0.3/magnum'
adding 'pymagnum-2.0.3/pymagnum.egg-info'
adding 'pymagnum-2.0.3/README.rst'
adding 'pymagnum-2.0.3/PKG-INFO'
adding 'pymagnum-2.0.3/setup.cfg'
adding 'pymagnum-2.0.3/setup.py'
adding 'pymagnum-2.0.3/magnum/__pycache__'
adding 'pymagnum-2.0.3/magnum/inverterdevice.py'
adding 'pymagnum-2.0.3/magnum/rtrdevice.py'
adding 'pymagnum-2.0.3/magnum/pt100device.py'
adding 'pymagnum-2.0.3/magnum/bmkdevice.py'
adding 'pymagnum-2.0.3/magnum/magparser.py'
adding 'pymagnum-2.0.3/magnum/agsdevice.py'
adding 'pymagnum-2.0.3/magnum/remotedevice.py'
adding 'pymagnum-2.0.3/magnum/aclddevice.py'
adding 'pymagnum-2.0.3/magnum/magtest.py'
adding 'pymagnum-2.0.3/magnum/__init__.py'
adding 'pymagnum-2.0.3/magnum/magnum.py'
adding 'pymagnum-2.0.3/magnum/magdump.py'
adding 'pymagnum-2.0.3/magnum/write_packets.py'
adding 'pymagnum-2.0.3/magnum/__pycache__/magtest.cpython-38.pyc'
adding 'pymagnum-2.0.3/magnum/__pycache__/magnum.cpython-38.pyc'
adding 'pymagnum-2.0.3/pymagnum.egg-info/requires.txt'
adding 'pymagnum-2.0.3/pymagnum.egg-info/dependency_links.txt'
adding 'pymagnum-2.0.3/pymagnum.egg-info/PKG-INFO'
adding 'pymagnum-2.0.3/pymagnum.egg-info/entry_points.txt'
adding 'pymagnum-2.0.3/pymagnum.egg-info/top_level.txt'
adding 'pymagnum-2.0.3/pymagnum.egg-info/SOURCES.txt'
Traceback (most recent call last):
  File "/usr/lib/python3.8/distutils/archive_util.py", line 237, in make_archive
    format_info = ARCHIVE_FORMATS[format]
KeyError: ''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "setup.py", line 15, in <module>
    setup(name='pymagnum',
  File "/usr/lib/python3.8/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib/python3.8/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/usr/lib/python3.8/site-packages/setuptools/command/sdist.py", line 54, in run
    self.make_distribution()
  File "/usr/lib/python3.8/site-packages/setuptools/command/sdist.py", line 78, in make_distribution
    orig.sdist.make_distribution(self)
  File "/usr/lib/python3.8/distutils/command/sdist.py", line 480, in make_distribution
    file = self.make_archive(base_name, fmt, base_dir=base_dir,
  File "/usr/lib/python3.8/distutils/cmd.py", line 369, in make_archive
    return archive_util.make_archive(base_name, format, root_dir, base_dir,
  File "/usr/lib/python3.8/distutils/archive_util.py", line 239, in make_archive
    raise ValueError("unknown archive format '%s'" % format)
ValueError: unknown archive format ''

**Comment by CharlesGodwin** [comment]
I don't know if you've read `how_to_build.md` but it describes some pre-requisites for the build that are not needed for an install. I improved some documentation and updated GitHub. I also discovered that build_draft.sh had DOS line endings, that has been fixed, although git may have autocorrected.

I assume you are using packet files from the `testdata` folder. I suggest you make a custom version that only contains packets for Magnum devices you actually have attached to your system. I have no idea what those are.

Have you looked in the examples folder? There may be something that you can easily use with Victron. Again, I don't know.

`magdump` not working surprises me. Send more details.

A quick thought! You don't have to build on your Venus OS machine, just copy the built zip file over and install it.

Thanks for your patience.

**Comment by tgsail1** [comment]
Thanks and yes. As it turned out I botched the repository download (ended up a level) last time. This time it loaded correctly- no errors. Thanks
T

**Comment by tgsail1** [comment]
Charles- Just to give you an update. I kept having trouble with the attempt at getting the two adapters to pass data reliably. I took one of the RPis down to the boat, connected an amazon-purchased telephone tee and connected the remote unit. One should not underestimate the general crappiness of stuff bought on amazon. Both splitters failed initially at running the remote unit and required a fair amount of pressure to get them to connect. Once connected, I was able to reliably get data off the bus (I’ve attached clips below). What was maybe not so surprising was that the remote got a bit flaky. I have to assume due to the crappy tee. Thanks for all your help. The magdump program has everything I need to import the data into the victron data bus. When done, I’ll post on the victron/venus GitHub site. Thanks T
data clips

***@***.***:~# magtest --device /dev/ttyUSB0 --timeout 0.01
Magnum Test Version:2.0.3
Options:cleanpackets=True, device='/dev/ttyUSB0', log=False, packets=50, timeout=0.01, trace=False
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500003B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500003B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500013B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500013B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500003B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500003B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000077B500013B18272023000000025800
Length:21 REMOTE_C0 =>000528905014280600649B86010F000000080000C0
Length:21 INVERTER  =>00000086000077B500013B18272023000000025800
Length:21 REMOTE_A0 =>000528905014280600649B86010F001614007300A0
Length:21 INVERTER  =>00000086000077B500003B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000077B500003B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000077B500013B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000077B500013B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:26 UNKNOWN   =>C21E0AF91202129218846212C30102C21E846254C30912001E0A
Length:10 UNKNOWN   =>000000000000E0E000F0
Length:21 INVERTER  =>00000085000077B500003B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500013B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500013B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500003B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500003B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500013B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000085000075B500013B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000085000075B500003B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500003B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000086000075B500013B18272023000000025800
Length:21 REMOTE_00 =>000528905014280600649B86010F00000000000000
Length:21 INVERTER  =>00000085000075B500013B18272023000000025800
Packets:48 of 50 with 2 UNKNOWN, in 15.18 seconds
INVERTER Detected
REMOTE Detected
RTR not connected
BMK not supported
AGS not connected
PT100 not connected
ACLD not supported




***@***.***:~# magdump --device /dev/ttyUSB0 --timeout .01
{"datetime":"2023-10-20T20:29:16+00:00","device":"MAGNUM","comm_device":"/dev/ttyUSB0","data":[{"device":"INVERTER","data":{"revision":"5.9","mode":64,"mode_text":"INVERT","fault":0,"fault_text":"None","vdc":12.2,"adc":5.0,"VACout":118.0,"VACin":0.0,"invled":1,"invled_text":"On","chgled":0,"chgled_text":"Off","bat":23.0,"tfmr":39.0,"fet":34.0,"model":35,"model_text":"MS2012","stackmode":0,"stackmode_text":"Stand Alone","AACin":0.0,"AACout":0.0,"Hz":60.0}},{"device":"REMOTE","data":{"revision":4.0,"searchwatts":5,"batterysize":400,"battype":0,"absorb":14.4,"chargeramps":80,"ainput":20,"parallel":60,"lbco":10.0,"vaccutout":155.0,"vsfloat":13.4,"vEQ":14.5,"absorbtime":1.5}}]}



***@***.***:~# magdump --device /dev/ttyUSB0 --timeout .01
/dev/ttyUSB0 There doesn't seem to be a network
Error: There are no usable devices connected.
***@***.***:~# magdump --device /dev/ttyUSB0 --timeout .01
{"datetime":"2023-10-20T20:33:09+00:00","device":"MAGNUM","comm_device":"/dev/ttyUSB0","data":[{"device":"INVERTER","data":{"revision":"5.9","mode":64,"mode_text":"INVERT","fault":0,"fault_text":"None","vdc":12.3,"adc":5.0,"VACout":120.0,"VACin":0.0,"invled":1,"invled_text":"On","chgled":0,"chgled_text":"Off","bat":24.0,"tfmr":39.0,"fet":35.0,"model":35,"model_text":"MS2012","stackmode":0,"stackmode_text":"Stand Alone","AACin":0.0,"AACout":0.0,"Hz":60.0}},{"device":"REMOTE","data":{"revision":4.0,"searchwatts":5,"batterysize":400,"battype":0,"absorb":14.4,"chargeramps":80,"ainput":20,"parallel":60,"lbco":10.0,"vaccutout":155.0,"vsfloat":13.4,"vEQ":14.5,"absorbtime":1.5}}]}
***@***.***:~# magdump --device /dev/ttyUSB0 --timeout .01
{"datetime":"2023-10-20T20:33:44+00:00","device":"MAGNUM","comm_device":"/dev/ttyUSB0","data":[{"device":"INVERTER","data":{"revision":"5.9","mode":64,"mode_text":"INVERT","fault":0,"fault_text":"None","vdc":12.3,"adc":5.0,"VACout":120.0,"VACin":0.0,"invled":1,"invled_text":"On","chgled":0,"chgled_text":"Off","bat":24.0,"tfmr":39.0,"fet":34.0,"model":35,"model_text":"MS2012","stackmode":0,"stackmode_text":"Stand Alone","AACin":0.0,"AACout":0.0,"Hz":60.0}},{"device":"REMOTE","data":{"revision":4.0,"searchwatts":5,"batterysize":400,"battype":0,"absorb":14.4,"chargeramps":80,"ainput":20,"parallel":60,"lbco":10.0,"vaccutout":155.0,"vsfloat":13.4,"vEQ":14.5,"absorbtime":1.5}}]}
***@***.***:~#

**Reply by CharlesGodwin** [reply]
Thanks for the update. I'm confused about your use of 2 raspberry Pi and dongles. Do you have two magnum systems? Or is it something obvious I missed.

Concerning the RS485 dongles. I have mush better success using the [RS485 HAT](https://www.waveshare.com/rs485-can-hat.htm) mentioned in the documentation.

**Comment by tgsail1** [comment]
No. I was using two RPis to talk to each other at home. One was running your Write_packets.py and the other magtest.py. The results were flakey. The RS485 dongles bought from amazon had very obvious counterfeit chips for the CH341 (fully unmarked). Either way the unit running magtest/magdum worked considerably better on the actual inverter network. I have an RS485 pi hat, but your comment about not being able to use bluetooth moved me towards USB. I need bluetooth for my install because I have temperature monitors on the boat using bluetooth LE.

**Reply by CharlesGodwin** [reply]
Ok. I understand now 

I'm glad it's working as hoped.

I'll publish the 2.0.3 version Sunday as the official release
